File: zadanie1.1/c/client/udp_client.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netdb.h>

int resolve_host(const char *host, int port, struct sockaddr_in *addr) {
    struct addrinfo hints, *res;

    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_DGRAM;

    if (getaddrinfo(host, NULL, &hints, &res) != 0) {
        perror("Błąd podczas rozwiązywania nazwy hosta");
        return -1;
    }

    memcpy(addr, res->ai_addr, sizeof(struct sockaddr_in));
    addr->sin_port = htons(port);

    freeaddrinfo(res);
    return 0;
}

void start_client(const char *host, int port) {
    int client_socket;
    struct sockaddr_in server_addr;
    socklen_t addr_len = sizeof(server_addr);
    int sizes[] = {1, 2, 100, 200, 500, 1000, 2000, 4000, 4095, 4096, 4097};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);
    unsigned char message[4096];

    if ((client_socket = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
        perror("Błąd tworzenia gniazda");
        exit(EXIT_FAILURE);
    }

    if (resolve_host(host, port, &server_addr) < 0) {
        fprintf(stderr, "Błąd rozwiązywania nazwy hosta: %s\n", host);
        close(client_socket);
        exit(EXIT_FAILURE);
    }

    for (int i = 0; i < num_sizes; i++) {
        int size = sizes[i];
        int content_length = size > 2 ? size - 2 : 0;

        message[0] = (content_length >> 8) & 0xFF;
        message[1] = content_length & 0xFF;

        for (int j = 0; j < content_length; j++) {
            message[2 + j] = 'A' + (j % 26);
        }

        printf("Wysyłanie wiadomości - rozmiar zawartości %d bajtów...\n", size);
        sendto(client_socket, message, size, 0, (const struct sockaddr *)&server_addr, addr_len);

        struct timeval timeout = {1, 0};
        setsockopt(client_socket, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));

        char response[128];
        int response_len = recvfrom(client_socket, response, sizeof(response), 0, (struct sockaddr *)&server_addr, &addr_len);

        if (response_len > 0) {
            response[response_len] = '\0';
            if (strcmp(response, "OK") == 0) {
                printf("Otrzymano potwierdzenie dla datagramu o rozmiarze %d bajtów\n", size);
            } else {
                printf("Odpowiedź: %s\n", response);
            }
        } else {
            printf("Brak odpowiedzi\n");

        }

        sleep(1);
    }

    close(client_socket);
}

int main(int argc, char *argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Użycie: %s <host> <port>\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    const char *host = argv[1];
    int port = atoi(argv[2]);

    start_client(host, port);
    return 0;
}

File: zadanie1.1/python/client/udp_client.py
import socket
import time
import sys


def start_client(host='localhost', port=12345):
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sizes = [1, 2, 100, 200, 1000, 2000] # tablica rozmiarów przesyłanych wiaodmości
    # sizes = list(range(4090, 4100, 1)) # nie zadziała poprawnie dla wartości większych niż 4096 - bo taki przykładowy rozmiar bufora w serwerze
    # sizes = list(range(65500, 65510)) # nie zostanie obsłużone - bo system nie obsługuje datagramów powyżej 65507 bajtów

    for size in sizes:
        length = (size).to_bytes(2, byteorder='big') # 2 pierwsze bajty na rozmiar treści
        content = bytes([(65 + (i % 26)) for i in range(size)]) # stworzenie wiadomości
        message = length + content

        print(f"Wysyłanie wiadomości - rozmiar wiadomości {size}B (całego datagramu {size+2}B)...")
        client_socket.sendto(message, (host, port))

        try:
            client_socket.settimeout(1)
            response, _ = client_socket.recvfrom(128)
            if response == b"OK":
                print(f"Otrzymano potwierdzenie dla wiadomości o rozmiarze {size}B (całego datagramu {size+2}B)")
            else:
                print(response.decode('utf-8'))
        except socket.timeout:
            print("Brak odpowiedzi")
            break
        print()

        time.sleep(1)

if __name__ == "__main__":
    start_client(sys.argv[1], int(sys.argv[2]))
